File: ios/Samples/Showcase-ObjectiveC/LiferayScreens-Showcase-ObjectiveC/CommentListScreenletViewController.m
//
//  CommentListScreenletViewController.m
//  LiferayScreens-Showcase-ObjectiveC
//
//  Created by Victor Galán on 03/02/2017.
//  Copyright © 2017 liferay. All rights reserved.
//

@import LiferayScreens;
#import "LiferayLogger.h"
#import "CommentListScreenletViewController.h"

@interface CommentListScreenletViewController () <CommentListScreenletDelegate>

@property (weak, nonatomic) IBOutlet CommentListScreenlet *listScreenlet;
@property (strong, nonatomic) CommentEditViewController_default *editViewController;

@end

@implementation CommentListScreenletViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    self.listScreenlet.delegate = self;
    self.listScreenlet.presentingViewController = self;
    self.listScreenlet.className = [LiferayServerContext stringPropertyForKey:@"commentScreenletClassName"];
    self.listScreenlet.classPK = [LiferayServerContext longPropertyForKey:@"commentScreenletClassPK"];

    self.definesPresentationContext = true;
}

- (IBAction)insertButtonPressed:(NSObject *) sender {
    if (!self.editViewController) {
        self.editViewController = [[CommentEditViewController_default alloc]initWithBody:@""];
        self.editViewController.modalPresentationStyle = UIModalPresentationOverCurrentContext;
        self.editViewController.confirmButton.titleLabel.text = @"Add comment";

        __weak typeof(self) weakSelf = self;
        self.editViewController.confirmBodyClosure = ^void (NSString *body) {
            [weakSelf.editViewController dismissViewControllerAnimated:YES completion:nil];

            if (body) {
                CommentAddInteractor *interactor = [[CommentAddInteractor alloc] initWithClassName:weakSelf.listScreenlet.className classPK:weakSelf.listScreenlet.classPK body:body];

                interactor.onSuccess = ^{
                    if (interactor.resultComment) {
                        [weakSelf.listScreenlet addComment:interactor.resultComment];
                    }
                };

                interactor.onFailure = ^(NSError *error) {
                    LiferayLog(error);
                };

                [interactor start];
            }
        };
    }

    [self presentViewController:self.editViewController animated:YES completion:nil];
}


- (void)screenlet:(CommentListScreenlet * _Nonnull)screenlet onListResponseComments:(NSArray<Comment *> * _Nonnull)comments {
    LiferayLog(comments);
}

- (void)screenlet:(CommentListScreenlet * _Nonnull)screenlet onCommentListError:(NSError * _Nonnull)error {
	LiferayLog(error);
}

- (void)screenlet:(CommentListScreenlet * _Nonnull)screenlet onSelectedComment:(Comment * _Nonnull)comment {
	LiferayLog(comment);
}

- (void)screenlet:(CommentListScreenlet * _Nonnull)screenlet onDeletedComment:(Comment * _Nonnull)comment {
	LiferayLog(comment);
}

- (void)screenlet:(CommentListScreenlet * _Nonnull)screenlet onCommentDelete:(Comment * _Nonnull)comment onError:(NSError * _Nonnull)error {
	LiferayLog(comment, error);
}

- (void)screenlet:(CommentListScreenlet * _Nonnull)screenlet onUpdatedComment:(Comment * _Nonnull)comment {
	LiferayLog(comment);
}

- (void)screenlet:(CommentListScreenlet * _Nonnull)screenlet onCommentUpdate:(Comment * _Nonnull)comment onError:(NSError * _Nonnull)error {

	LiferayLog(comment, error);
}


@end

File: ios/Samples/Showcase-swift/LiferayScreens-Showcase-Swift/ViewControllers/CommentListScreenletViewController.swift
/**
* Copyright (c) 2000-present Liferay, Inc. All rights reserved.
*
* This library is free software; you can redistribute it and/or modify it under
* the terms of the GNU Lesser General Public License as published by the Free
* Software Foundation; either version 2.1 of the License, or (at your option)
* any later version.
*
* This library is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
* FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
* details.
*/
import UIKit
import LiferayScreens


class CommentListScreenletViewController: UIViewController,
	CommentListScreenletDelegate {

	@IBOutlet weak var listScreenlet: CommentListScreenlet?

	var editViewController: CommentEditViewController_default?

	override func viewDidLoad() {
		super.viewDidLoad()

		self.definesPresentationContext = true

		self.listScreenlet?.delegate = self
		self.listScreenlet?.presentingViewController = self
		self.listScreenlet?.loadList()
	}

	@IBAction func insertButtonPressed(sender: AnyObject) {
		if editViewController == nil {
			editViewController = CommentEditViewController_default(body: "")
			editViewController!.modalPresentationStyle = .OverCurrentContext
			editViewController!.confirmButton?.titleLabel?.text = "Add comment"
			editViewController!.confirmBodyClosure = addComment
			presentViewController(editViewController!, animated: true, completion: nil)
		}
	}

	//MARK: CommentListScreenletDelegate

	func screenlet(screenlet: CommentListScreenlet, onCommentListError error: NSError) {
		print("DELEGATE: onCommentListError called -> \(error)\n")
	}

	func screenlet(screenlet: CommentListScreenlet, onSelectedComment comment: Comment) {
		print("DELEGATE: onCommentSelected called -> \(comment)\n")
	}

	func screenlet(screenlet: CommentListScreenlet, onListResponseComments comments: [Comment]) {
		print("DELEGATE: onCommentListResponse called -> \(comments)\n")
	}

	func screenlet(screenlet: CommentListScreenlet, onDeletedComment comment: Comment) {
		print("DELEGATE: onDeletedComment called -> \(comment)\n")
	}

	func screenlet(screenlet: CommentListScreenlet, onCommentDelete comment: Comment,
			onError error: NSError) {
		print("DELEGATE: onCommentDelete onError called -> \(comment) \(error)\n")
	}

	func screenlet(screenlet: CommentListScreenlet, onUpdatedComment comment: Comment) {
		print("DELEGATE: onUpdatedComment called -> \(comment)\n")
	}

	func screenlet(screenlet: CommentListScreenlet, onCommentUpdate comment: Comment,
			onError error: NSError) {
		print("DELEGATE: onCommentUpdate onError called -> \(comment) \(error)\n")
	}

	//MARK: Private methods

	private func addComment(body: String?) {
		editViewController?.dismissViewControllerAnimated(true) {
			self.editViewController = nil
		}

		if let newCommentBody = body, screenlet = listScreenlet {
			let interactor = CommentAddInteractor(
				screenlet: nil,
				groupId: screenlet.groupId,
				className: screenlet.className,
				classPK: screenlet.classPK,
				body: newCommentBody)

			interactor.onSuccess = {
				if let comment = interactor.resultComment {
					screenlet.addComment(comment)
				}
			}

			interactor.onFailure = {
				print("ERROR: adding comment \($0)")
			}

			interactor.start()
		}
	}

}
